src/Avalonia.Styling/LogicalTree/ILogicalRoot.cs

```diff
@@ -0,0 +1,9 @@
+﻿namespace Avalonia.LogicalTree
+{
+    /// <summary>
+    /// Represents a root of a logical tree.
+    /// </summary>
+    public interface ILogicalRoot : ILogical
```

Comment:
What prevents us from using `ILogical` directly instead of having `ILogicalRoot`? `ILogicalRoot` adds nothing.

Reply:
It acts as a marker for the end of the logical tree. When traversing up the tree, you continue until you find an `ILogicalRoot`. If none is found, then the tree is unrooted.

Reply:
What's the difference between having a root object and some object that has no parent? Is it just not attached if it has no parent?

Reply:
Not 100% sure what you mean, but consider a `StackPanel` with a `Button` child, where the `StackPanel` has no parent:

```
StackPanel
  |- Button
```

The `Button` here has a parent, but the `StackPanel` does not. Because there's no logical root in this case, both the `Button` and the `StackPanel` are detached from a logical tree. They can't be shown on-screen, they won't have styles applied, they won't be inheriting a `DataContext` etc etc. It's only when the `StackPanel` gets attached to a logical tree that these things start working.

Reply:
In my understanding, it is enough to walk up the logical tree until the parent is null. The root object has nothing special it is just some interface you are looking for. As you described above once a control is attached to the logical tree it can traverse up the logical tree up until it reaches an `ILogical` that has no parent. That indicates the end.

Reply:
No, read what I said again, I'm not sure how i can phrase it differently... Maybe some code will help:

```
var sp = new StackPanel();
var b = new Button();
sp.Children.Add(b);
```

The `Parent` of `sp` is null. Is `b` attached to a logical tree now? No, it's not.

Reply:
What is the difference between a `TopLevel` instance and a `StackPanel` instance? Both would function a the root element of some tree. When you reach `ILogicalRoot` you will most likely cast it to a different type. Both scenarios build up a tree.

Reply:
Just something to consider: with our current architecture we can't have `Window` being a child to anything else, e. g. MDI window host or even our own "display" running as a VNC server root or directly on top of the OS kernel.

Reply:
> What is the difference between a TopLevel instance and a StackPanel instance?

Lets talk about a standalone `Button`:

- It has no renderer so cannot be displayed on-screen
- It has no toplevel bounds, therefore it can't be measured
- It probably won't have any styles, so no template
- It may be waiting for attachment to the logical tree for its `DataContext` so its bindings are probably non-functional
- It doesn't know its scaling setting
- ...etc etc

> Just something to consider: with our current architecture we can't have Window being a child to anything else, e. g. MDI window host or even our own "display" running as a VNC server root or directly on top of the OS kernel.

Yeah, we will need to consider how to do this when the need arises!

Reply:
I guess I have to accept this marker interface because of your explanations. It is probably the easiest way to implement this.